File: loader/Main/Plugins.cs
﻿using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace PenguLoader.Main
{
    public static class Plugins
    {
        public static List<PluginInfo> All()
        {
            var pluginsDir = Config.PluginsDir;
            var plugins = new List<PluginInfo>();

            // sub-folders plugins.
            foreach (var dir in Directory.GetDirectories(pluginsDir))
            {
                var dirName = Path.GetFileName(dir);
                if (!FilterName(dirName)) continue;
                var disabled = false;
                string indexPath;

                if (File.Exists(indexPath = Path.Combine(dir, "index.js")) ||
                    (disabled = File.Exists(indexPath = Path.Combine(dir, "index.js_"))))
                    AddPlugin(dirName, indexPath, !disabled);
            }

            // top-level plugins.
            foreach (var path in Directory.GetFiles(pluginsDir, "*.*", SearchOption.TopDirectoryOnly))
            {
                var fileName = Path.GetFileName(path);
                var disabled = false;

                if (FilterName(fileName) &&
                    (fileName.EndsWith(".js") || (disabled = fileName.EndsWith(".js_"))))
                    AddPlugin(fileName.TrimEnd('_'), path, !disabled);
            }

            return plugins;

            void AddPlugin(string name, string path, bool enabled)
            {
                var plugin = new PluginInfo
                {
                    Name = name,
                    Path = path,
                    Enabled = enabled
                };

                ParsePluginEntry(path, plugin);
                plugins.Add(plugin);
            }
        }

        public static void Toggle(PluginInfo plugin)
        {
            var path = plugin.Path;

            if (File.Exists(path))
            {
                if (path.EndsWith(".js_"))
                {
                    File.Move(path, plugin.Path = path.TrimEnd('_'));
                    plugin.Enabled = true;
                }
                else if (path.EndsWith(".js"))
                {
                    File.Move(path, plugin.Path = path + '_');
                    plugin.Enabled = false;
                }
            }
        }

        // Ignore name starts with . or _
        private static bool FilterName(string name)
        {
            return !(name.StartsWith(".") || name.StartsWith("_"));
        }

        // Get @author and @link.
        private static void ParsePluginEntry(string path, PluginInfo plugin)
        {
            if (!File.Exists(path)) return;
            var content = File.ReadAllText(path, Encoding.UTF8);

            var author = GetTagValue(content, "author");
            if (!string.IsNullOrEmpty(author))
                plugin.Author = author.Contains("#") ? author : '@' + author;

            var link = GetTagValue(content, "link");
            if (link.StartsWith("http://") || link.StartsWith("https://"))
                plugin.Link = link;
        }

        // Parse @tag in jsdoc
        private static string GetTagValue(string jsdoc, string tagName)
        {
            var pattern = $"@{tagName}\\s+(.+)";
            var match = Regex.Match(jsdoc, pattern);
            return match.Success ? match.Groups[1].Value.Trim() : string.Empty;
        }

        public class PluginInfo
        {
            public string Author = string.Empty;
            public bool Enabled;
            public string Link = string.Empty;
            public string Name;
            public string Path;
        }
    }
}

File: loader_old/src/lib/plugins.ts
import { join } from '@tauri-apps/api/path'
import { exists, readTextFile, readDir, createDir } from '@tauri-apps/api/fs'
import { Shell } from './shell'
import { Config } from './config'
import { getHash } from './utils'

export interface PluginInfo {
  name: string
  description?: string
  author?: string
  link?: string

  path: string
  entryPath: string
  hash: number
}

export const PluginManager = new class {

  private disabledSet = new Set<number>()

  private getDir() {
    let path = Config.get('app', 'plugins_dir', '')
    if (!path || path.startsWith('.')) {
      path = Config.basePath('plugins')
    }
    return path
  }

  async openFolder() {
    const dir = this.getDir()
    if (!await exists(dir)) {
      await createDir(dir, {
        recursive: true
      })
    }
    await Shell.expandFolder(dir)
  }

  /**
   * Get all plugins.
   */
  async getPlugins() {
    const dir = this.getDir()
    const plugins = Array<PluginInfo>()

    this.disabledSet = this.fetchDisabledSet()

    function push(name: string, _path: string, entry: string) {
      const dir2 = dir.replace(/\\/g, '/')
      const url = entry.replace(/\\/g, '/')

      const shortPath = url.replace(dir2, '').substring(1)
      const hash = getHash(shortPath.toLowerCase())

      plugins.push({
        name: name,
        path: shortPath,
        entryPath: entry,
        hash: hash,
      })
    }

    if (await exists(dir)) {
      const ref = { entry: '' }
      for (const file of await readDir(dir)) {
        if (file.children) {
          // scan @author folder
          if (file.name!.startsWith('@')) {
            for (let subdir of await readDir(file.path)) {
              // subfolder plugin that contains index.js
              if (subdir.children
                && this.allowedName(subdir.name)
                && await this.hasIndex(subdir.path, ref)) {
                const name = `${file.name}/${subdir.name}`
                push(name, subdir.path, ref.entry)
              }
            }
          }
          // subfolder plugin that contains index.js
          else if (this.allowedName(file.name) && await this.hasIndex(file.path, ref)) {
            push(file.name!, file.path, ref.entry)
          }
        }
        // top-level file plugin
        else if (this.allowedName(file.name) && await this.isIndex(file.path)) {
          const name = file.name!.substring(0, file.name!.lastIndexOf('.'))
          push(name, file.path, file.path)
        }
      }

      // parse metadata
      for (const plugin of plugins) {
        await this.parsePluginEntry(plugin)
      }
    }

    return plugins
  }

  isEnabled(hash: number) {
    return !this.disabledSet.has(hash)
  }

  async toggleState(hash: number) {
    if (this.disabledSet.has(hash)) {
      this.disabledSet.delete(hash)
    } else {
      this.disabledSet.add(hash)
    }

    const value = [...this.disabledSet].map(x => x.toString(16)).join()
    Config.set('app', 'disabled_plugins', value)
    await Config.save()

    return this.isEnabled(hash)
  }

  private allowedName(name?: string) {
    return typeof name === 'string'
      && !name.startsWith('_')
      && !name.startsWith('.')
  }

  private async isIndex(path: string, ref?: { entry: string }) {
    if (path.endsWith('.js') && await exists(path) || await exists(path += '_')) {
      if (typeof ref === 'object') {
        ref.entry = path
      }
      return true
    }
    return false
  }

  private async hasIndex(dir: string, ref: { entry: string }) {
    const path = await join(dir, 'index.js')
    return await this.isIndex(path, ref)
  }

  private async parsePluginEntry(plugin: PluginInfo) {
    if (await exists(plugin.entryPath)) {
      const content = await readTextFile(plugin.entryPath)
      const description = this.getTagValue(content, 'description')
      const author = this.getTagValue(content, 'author')
      const link = this.getTagValue(content, 'link')

      if (description)
        plugin.description = description

      if (author)
        plugin.author = author.includes('#') ? author : '@' + author

      if (link.startsWith('https://'))
        plugin.link = link
    }
  }

  // Parse @tag in jsdoc
  private getTagValue(jsdoc: string, tag: string) {
    const regex = new RegExp(`@${tag}\\s+(.+)`)
    const match = regex.exec(jsdoc)
    if (match) {
      return match[1].trim()
    }
    return ''
  }

  private fetchDisabledSet() {
    const set = new Set<number>()
    const rawSet = <string>Config.get('app', 'disabled_plugins', '')
    const hashes = rawSet.split(',')

    for (const hash of hashes) {
      const num = parseInt(hash.trim(), 16)
      if (num) set.add(num)
    }

    return set
  }
}